src/System.Private.CoreLib/shared/System/SpanHelpers.Byte.cs

```diff
@@ -959,25 +960,34 @@ public static unsafe int LastIndexOfAny(ref byte searchSpace, byte value0, byte
         public static unsafe bool SequenceEqual(ref byte first, ref byte second, nuint length)
         {
             if (Unsafe.AreSame(ref first, ref second))
+            {
                 goto Equal;
+            }
 
+            bool result = false;
             IntPtr i = (IntPtr)0; // Use IntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
             IntPtr n = (IntPtr)(void*)length;
 
-            if (Vector.IsHardwareAccelerated && (byte*)n >= (byte*)Vector<byte>.Count)
+            // Conditinally jump forward for Vector versions to give the non-Vector variant used for shorter
+            // lengths an unprimed branch prediction boost
+            if (Avx2.IsSupported && 
+                (byte*)n >= (byte*)Vector256<byte>.Count)
```

Comment:
Does this take into account the cases where the benefits of using `Vector256` outweigh the drawbacks (it doesn't look like it)?

Some examples are:
* For unaligned data, reads/writes crossing a cache-line boundary will happen twice as frequently (same with crossing page boundaries, for large enough data)
* For unaligned data, 32-byte reads are generally less efficient than 2x16-byte reads
* Additional saving/restoring of the upper 128-bits across method call boundaries
* Additional `vzeroupper` calls
* Possible frequency downscaling when executing a "heavy" 256-bit workload (this one tends to make micro-benches look good, but real world scenarios can actually regress)

Reply:
>  it is worthwhile to try to align the data to 16 or 32-byte boundaries by scalar operations before executing AVX2/SSE2 code.

Ah, this function has two input data. That is not possible to do such aligning. Please ignore that comment.

Reply:
Hmmm, interesting I didn't know that. It looks like the optimization manual doesn't indicate the second condition.

> 12.6.2 Consider 16-Byte Memory Access when Memory is Unaligned
>
> For best results use Intel AVX 32-byte loads and align data to 32-bytes. However, there are cases where
you cannot align the data, or data alignment is unknown. This can happen when you are writing a library
function and the input data alignment is unknown. In these cases, using 16-Byte memory accesses may
be the best alternative. The following method uses 16-byte loads while still benefiting from the 32-byte
YMM registers.
>
> Consider replacing unaligned 32-byte memory accesses using a combination of VMOVUPS,
VINSERTF128, and VEXTRACTF128 instructions.

Reply:
Yes, the manual is not very clear. If the data is not aligned to 16-bye, programs will get same numbers of cache-line splits on the same size of data comparing to 32-byte loads. Meanwhile, 16-bye loads for 32-byte computing has additional code size overhead.

> Aligning data to vector length is recommended. When using 16-byte SIMD instructions, loaded data
should be aligned to 16 bytes. Similarly, for best results when using Intel AVX instructions with 32-byte
registers align the data to 32-bytes.

Reply:
> Possible frequency downscaling when executing a "heavy" 256-bit workload

Seems to more heavily effect AVX-512 than AVX2 https://en.wikichip.org/wiki/intel/xeon_silver/4116#Frequencies
![image](https://user-images.githubusercontent.com/1142958/51279297-e78c3f00-19d4-11e9-8001-f36dd2549ba5.png)

Other issues exist in current code; though obv would be helpful if I can improve some CQ at the same time

Reply:
> Seems to more heavily effect AVX-512 than AVX2 

Right, we only need to consider the frequency down on AVX-512. That is not a big issue on AVX/AVX2.

Reply:
Yeah, a 48% drop is a lot worse, but a 19% drop in cpu frequency is still pretty bad as well and should be taken into account here (since it will impact other workloads that run immediately after a method that causes the drop and until the frequency recovers to normal levels).

Reply:
From my experience, we never came cross perf regression due to AVX/AVX2 frequency drop (but we indeed detected considerable frequency issues from AVX-512). Meanwhile, the trigger condition is non-trivial, only a few AVX/AVX2 instruction execution won't make measurable frequency drop.

Reply:
4% on 1 active core (3GHz -> 2.9GHz) and 12.5% on 12 active cores (2.4Ghz -> 2.1 GHz)?

We already know this is a benefit beyond "micro-benches" as Bing is running the AVX2 version; and they [reported it as their number 1 improvement in performance](https://blogs.msdn.microsoft.com/dotnet/2018/08/20/bing-com-runs-on-net-core-2-1/)

This change introduces SSE2 for shorter lengths when AVX2 is available; rather than only working on 32 bytes or greater; it now also works on 16 bytes or larger.

Reply:
> This change introduces SSE2 for shorter lengths when AVX2 is available; rather than only working on 32 bytes or greater; it now also works on 16 bytes or larger.

Right, and I have no doubt Avx2 is beneficial, especially on larger workloads.

However, I would expect the cutoff points to be different. At least based on prior experience, 128-bit code paths tend to be better than 256-bit code paths until the total bytes processed are around 64-128 bytes, at which point Avx2 starts taking over.

There are of course various other things to take into account here, such as alignment, whether any trailing elements can still be processed using vectorization, etc. But I wouldn't expect Avx2 to be faster than sse just starting at 32-bytes.

Reply:
Changed Avx2 to `>= Vector256<byte>.Count * 2` and started the method with an unconditional 2x unroll

Added 2x unroll to Sse2; which is conditional based on length